utils.py

```diff
@@ -269,32 +266,41 @@ def uploadDebArtifacts():
     )
 
 
-def staging_branch_fn(branch):
+def staging_branch_fn(branch: str) -> bool:
     return fnmatch.fnmatch(branch, "prot-st-*")
 
 
-def fnmatch_any(s, list_of_patterns):
-    return any(fnmatch.fnmatch(s, p) for p in list_of_patterns)
+def fnmatch_any(branch: str, patterns: list[str]) -> bool:
+    return any(fnmatch.fnmatch(branch, pattern) for pattern in patterns)
 
 
 # Priority filter based on saved package branches
-def nextBuild(bldr, requests):
-    for r in requests:
-        if fnmatch_any(r.sources[""].branch, releaseBranches):
-            return r
-    for r in requests:
-        if fnmatch_any(r.sources[""].branch, savedPackageBranches):
-            return r
-    return requests[0]
-
-
-def canStartBuild(builder, wfb, request):
-    worker = wfb.worker
-    if not "s390x" in worker.name:
+def nextBuild(builder: Builder, requests: list[BuildRequest]) -> str:
+    def build_request_sort_key(request: BuildRequest):
+        branch = request.sources[""].branch
+        # Booleans are sorted False first.
+        # Priority is given to releaseBranches, savePackageBranches
+        # then it's first come, first serve.
+        return (
+            not fnmatch_any(branch, releaseBranches),
+            not fnmatch_any(branch, savedPackageBranches),
+            request.getSubmitTime(),
```

Comment:
I need to check in BB code, maybe you already did that, but are you sure that the `requests` object doesn't have the **oldest build request** at `[0]` ? If yes, then sorting is not needed, FIFO being satisfied by design.

Reply:
More on that, we should first parse `requests`,
and return the **most recent request** for a match on **releaseBranches**.

For example, let's say we have the following case:
-> request 1 - `branch bb-10.5-release` - tarball **1**
-> request 2 - `branch bb-10.5-release` - tarball **2**

It's obvious that we're interested in tarball 2 to finish first because it potentially contains a fix that makes it the candidate for release.